glean/src/server_events.rs

```diff
@@ -0,0 +1,308 @@
+            ip_address: Some(config.ip_address.clone()),
+            payload: serde_json::to_string(payload).expect("unable to marshal payload to json."),
+        }
+    }
+
+    /// Method called by each ping-specific record method.
+    /// The goal is to construct the ping, wrap it in the envelope
+    /// and print to stdout.
+    fn record(
+        &self,
+        document_type: &str,
+        request_info: &RequestInfo,
+        metrics: Metrics,
+        events: Vec<GleanEvent>,
+    ) {
+        let telemetry_payload: PingPayload = PingPayload {
+            client_info: self.create_client_info(),
+            ping_info: PingInfo::default(),
+            metrics,
+            events,
+        };
+
+        let ping: Ping = self.create_ping(document_type, request_info, &telemetry_payload);
+
+        let envelope: LogEnvelope = LogEnvelope {
+            log_type: GLEAN_EVENT_MOZLOG_TYPE.to_owned(),
+            fields: ping,
+        };
+        let envelope_json =
+            serde_json::to_string(&envelope).expect("unable to marshal payload to json.");
+        println!("{}", envelope_json);
+    }
+}
+
+// Code below is generated based on the provided `metrics.yaml` file:
+
+// Metrics of the `event` type. Anything defined in `extra_keys` has it's own struct field.
+// The appended `Event` term to any metric of the event type implies the ping event.
+
+pub struct SyncstorageGetCollectionsEvent {
+    // metadata for event in `extra_keys`
+}
+
+// Implementing the EventsPingEvent trait for the generated struct SyncstorageGetCollectionsEvent
+impl EventsPingEvent for SyncstorageGetCollectionsEvent {
+    /// Create a GleanEvent for the above-defined Event struct (SyncstorageGetCollectionsEvent).
+    /// Any metadata `extra` values are passed into the extra HashMap.
+    fn glean_event(&self) -> GleanEvent {
+        // Any `extra_keys` will be output below to be inserted into `extra`.
+        // If there are none, an empty, immutable HashMap is created.
+        let extra: HashMap<String, String> = HashMap::new();
+
+        new_glean_event("syncstorage", "get_collections", extra)
+    }
+}
+
+/// Marker trait for events per ping
+pub trait EventsPingEvent {
+    fn glean_event(&self) -> GleanEvent; // Returns an instance of GleanEvent
+}
+
+pub struct EventsPing {
+    pub syncstorage_device_family: String, // Device family from which sync action was initiated. Desktop PC, Tablet, Mobile, and Other.
+    pub syncstorage_hashed_device_id: String, // Hashed device id that is associated with a given account.
+    pub syncstorage_hashed_fxa_uid: String, // User identifier. Uses `hashed_fxa_uid` for accurate count of sync actions.
+    pub syncstorage_platform: String, // Platform from which sync action was initiated. Firefox Desktop, Fenix, or Firefox iOS.
+    pub event: Option<Box<dyn EventsPingEvent>>, // valid event of `EventsPingEvent` for this ping.
+}
+
+// Record and submit `events` ping
```

Comment:
We should make this a doc string, since we've got one for the `record_events_ping()` function.